File: AquariumApi.Models/Device/DeviceScheduleTask.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace AquariumApi.Models
{
    [Table("tblDeviceTask")]
    public class DeviceScheduleTask
    {
        public int? Id { get; set; } = null;
        public string Name { get; set; }
        public List<DeviceAction> Actions { get; set; } = new List<DeviceAction>();
        public int? TriggerSensorId { get; set; }
        public GpioPinValue? TriggerSensorValue { get; set; }
        public int MaximumRuntime { get; set; }

        [ForeignKey("TriggerSensorId")]
        public virtual DeviceSensor TriggerSensor { get; set; }
        public ScheduleTaskTypes TaskTypeId { get; set; }
    }
}

File: AquariumApi.DeviceApi/ClientApp/src/app/modules/SharedDeviceModule/models/DeviceScheduleTask.ts
import { DeviceScheduleTaskTypes } from "../../../models/types/DeviceScheduleTaskTypes"
import { DeviceSensor } from "./DeviceSensor"
import { GpioPinValue } from "./GpioPinValue"
import { DeviceAction } from "./DeviceAction"
import { DeviceConfiguration } from "./DeviceConfiguration"

export class DeviceScheduleTask {
  id: number //= Math.floor(Math.random() * 100);
  name: string
  actions: DeviceAction[] = []
  triggerSensorId: number | null
  triggerSensor: DeviceSensor | null
  triggerSensorValue: GpioPinValue | null
  maximumRuntime: number


  static expandTask(task: DeviceScheduleTask, device: DeviceConfiguration):DeviceScheduleTask {
    var task = {...task}
    task.actions = task.actions.map(a => {
      if(isNaN(a.targetSensorId)) return {...a}
      return {
        ...a,
        targetSensor: device.sensors.filter(x => x.id == a.targetSensorId)[0]
      }
    })
    if(isNaN(task.triggerSensorId)) return {...task}
    return {
      ...task,
      triggerSensor: device.sensors.filter(x => x.id == task.triggerSensorId)[0]
    }
  }
  /*
  static getETAForTask(task: DeviceScheduleTask) {
    var d = moment(task.startTime).diff(moment());
    return moment.duration(d).humanize();
  }
  */
}

